New flattening impl

klkvr opened the pull request:
Ref https://github.com/foundry-rs/foundry/issues/4034.

## Solution
solang-parser AST was not sufficient to rename stuff, because it doesn't collect IDs of declarations and we can't match specific identifier to the declaration. Native solc AST gives such ability, so I've implemented native solc AST visitor.

The implementation of flattening itself consits of several steps:
1. Find all sources that are needed for target (imports of any depth).
2. Expore ASTs for all sources and find all top-level declarations (contracts, events, structs, etc).
3. Find if any top-level declarations names are same
4. If any duplicates are found, figure out new names for them (e.g. 2 `OracleLike` interfaces become `OracleLike_0` and `OracleLike_1`)
5. Walk AST and find all references to top-level declarations. Replace every reference with declaration name. We should update names even for unchanged declarations to deal with aliased imports.
6. Collect and remove all import directives.
7. Collect and remove all pragmas and license identifiers.

This flattener implementation can be a full replacer of the current impl for all cases when source being flattened can be compiled via solc.

klkvr commented:
This should also fix issues with aliases (like https://github.com/foundry-rs/foundry/issues/2545, https://github.com/foundry-rs/foundry/issues/3265). 

Current impl handles aliases renaming via regex'es and may often be inaccurate.

DaniPopes commented on src/flatten.rs:
```diff
@@ -0,0 +1,398 @@
+            ordered_deps: &mut Vec<PathBuf>,
+        ) -> Result<()> {
+            if deps.insert(path.clone()) {
+                let target_dir = path.parent().ok_or_else(|| {
+                    SolcError::msg(format!(
+                        "failed to get parent directory for \"{:?}\"",
+                        path.display()
+                    ))
+                })?;
+
+                let node_id = graph.files().get(path).ok_or_else(|| {
+                    SolcError::msg(format!("cannot resolve file at {:?}", path.display()))
+                })?;
```
use either `{}, path.display()` or `{:?}, path`

klkvr commented:
I've refactored code and added logic for correct processing of references to types declared in scope of contracts.

It requires such workaround because for some reason solc does not give any information about the parent contract for `UserDefinedTypeName` objects representing types like `ParentContract.EnumName`. Such node will only have AST ID and location of `EnumName` definition included.

That way, if we want to rename contract `ParentContract`, we need to be able to determine that AST ID of `EnumName` is related to that contract and replace it with `ParentContract_i.EnumName`

mattsse commented on src/flatten.rs:
```diff
@@ -0,0 +1,561 @@
+    /// 3. Remove all imports.
+    /// 4. Remove all pragmas except for the ones in the target file.
+    /// 5. Remove all license identifiers except for the one in the target file.
+    pub fn flatten(&self) -> String {
```
can we split this into a two step process,

one additional (internal) struct that wraps Updates + Flattener, basically `let updated_sources = self.get_updated_sources(updates)`

that then has function to construct the output string. this could perhaps come in handy for some tests

mattsse reviewed (changes requested):
this looks great!!!

can add some additional tests for #34 ?

mattsse commented:
@sakulstra I believe you had some issues with flattening in the past, do you know any projects we should test this against?

sakulstra commented:
@mattsse there were a few issues/things i commented in the past:
- https://github.com/foundry-rs/foundry/issues/2378
e.g. `cast etherscan-source -d etherscan/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0 0x7b2a3cf972c3193f26cdec6217d27379b6417bd0 && forge flatten ./etherscan/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0/AToken/@aave/protocol-v2/contracts/protocol/tokenization/AToken.sol --output ./etherscan/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0/Flattened.sol` would remove `pragma experimental ABIEncoderV2;` and therefore creating non-identical(in this case breaking) code.

- https://github.com/foundry-rs/foundry/issues/6539 not really an issue but more a feature request i guess?

For #4034 specifically we also had some issues with that, but can't recall right now in which contracts. Will check if i can find on monday.

klkvr commented:
Added test and fix for https://github.com/foundry-rs/foundry/issues/2378

klkvr commented:
Pushed solution for https://github.com/foundry-rs/foundry/issues/6539.

https://github.com/foundry-rs/compilers/blob/ce8e42746599694bb884f06e69d5014e5d7a10e8/src/flatten.rs#L586-L624

 I've also shared it's logic with current flattening implementation and rewritten it a bit as it's not longer need to be recursive.

mattsse commented on src/flatten.rs:
```diff
@@ -0,0 +1,623 @@
+
+impl ItemLocation {
+    fn try_from_source_loc(src: &SourceLocation, path: PathBuf) -> Option<Self> {
+        if src.start.is_none() || src.length.is_none() {
+            return None;
+        }
+        let start = src.start.unwrap();
+        let end = start + src.length.unwrap();
```
use `?` operator here

mattsse commented on src/flatten.rs:
```diff
@@ -0,0 +1,623 @@
+    /// 4. Remove all pragmas except for the ones in the target file.
+    /// 5. Remove all license identifiers except for the one in the target file.
+    pub fn flatten(&self) -> String {
+        let mut updates = Updates::new();
+
+        let top_level_names = self.rename_top_level_definitions(&mut updates);
+        self.rename_contract_level_types_references(&top_level_names, &mut updates);
+        self.remove_imports(&mut updates);
+        let target_pragmas = self.process_pragmas(&mut updates);
+        let target_license = self.process_licenses(&mut updates);
+
+        let result = FlatteningResult::new(self, updates, target_pragmas, target_license);
```
FlattenedResult only borrows from `Flattener`, right?

let's move this to `fn flatten_result` so that this function becomes `self.flatten_result().get_flattened_target`
if possible

mattsse reviewed (approved):
nice, last cosmetic nits!

klkvr commented:
@mattsse pushed fixes

let's merge a bit later, I will now write integration of it into foundry and may decide to change the API of `Flattener` a little bit

mattsse reviewed (approved):
sgtm!
pending foundry companion pr

Evalir commented on src/config.rs:
```diff
@@ -401,110 +402,95 @@ impl ProjectPathsConfig {
-
-        let mut imports = target_node.imports().clone();
-        imports.sort_by_key(|x| x.loc().start);
+        let ordered_deps = collect_ordered_deps(&flatten_target, self, &graph)?;
+
+        let mut sources = Vec::new();
+
+        let mut result = String::new();
+
+        for path in ordered_deps.iter() {
+            let node_id = graph.files().get(path).ok_or_else(|| {
+                SolcError::msg(format!("cannot resolve file at {}", path.display()))
+            })?;
+            let node = graph.node(*node_id);
+            let mut content = node.content().to_owned();
+
+            for alias in node.imports().iter().flat_map(|i| i.data().aliases()) {
+                let (alias, target) = match alias {
+                    SolImportAlias::Contract(alias, target) => (alias.clone(), target.clone()),
+                    _ => continue,
+                };
+                let name_regex = utils::create_contract_or_lib_name_regex(&alias);
+                let target_len = target.len() as isize;
+                let mut replace_offset = 0;
+                for cap in name_regex.captures_iter(&content.clone()) {
+                    if cap.name("ignore").is_some() {
+                        continue;
+                    }
+                    if let Some(name_match) =
+                        ["n1", "n2", "n3"].iter().find_map(|name| cap.name(name))
+                    {
+                        let name_match_range =
+                            utils::range_by_offset(&name_match.range(), replace_offset);
+                        replace_offset += target_len - (name_match_range.len() as isize);
+                        content.replace_range(name_match_range, &target);
+                    }
+                }
+            }
```
could we add a bit of docs over what's happening here?

klkvr replied:
will add, however this code was written before me so I might get something wrong haha

Evalir commented on src/flatten.rs:
```diff
@@ -0,0 +1,627 @@
+    utils, Graph, Project, ProjectCompileOutput, ProjectPathsConfig, Result,
+};
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
```
can we add some docs here?

Evalir reviewed (approved):
amazing. some nits on docs only
